draft-ietf-quic-tls.md

```diff
@@ -440,102 +467,63 @@ Important:
-response to this request, QUIC cannot ask for client address validation, it can
-only abort or permit the connection attempt to proceed.
-
-QUIC can provide a new address validation token for use in session resumption at
-any time after the handshake is complete.  Each time a new token is provided TLS
-generates a NewSessionTicket message, with the token included in the ticket.
-
-See {{client-address-validation}} for more details on client address validation.
-
-
-### Key Ready Events
-
-TLS provides QUIC with signals when 0-RTT and 1-RTT keys are ready for use.
+At each change of encryption level in either direction, TLS signals
+QUIC, providing the new level and the encryption keys. It will also
 These events are not asynchronous, they always occur immediately after TLS is
 provided with new handshake octets, or after TLS produces handshake octets.
 
-When TLS completed its handshake, 1-RTT keys can be provided to QUIC.  On both
-client and server, this occurs after sending the TLS Finished message.
-
-This ordering means that there could be frames that carry TLS handshake messages
-ready to send at the same time that application data is available.  An
-implementation MUST ensure that TLS handshake messages are always sent in
-packets protected with handshake keys (see {{handshake-secrets}}).  Separate
-packets are required for data that needs protection from 1-RTT keys.
-
 If 0-RTT is possible, it is ready after the client sends a TLS ClientHello
 message or the server receives that message.  After providing a QUIC client with
-the first handshake octets, the TLS stack might signal that 0-RTT keys are
-ready.  On the server, after receiving handshake octets that contain a
+the first handshake octets, the TLS stack might signal the change to the
+the 0-RTT keys. On the server, after receiving handshake octets that contain a
 ClientHello message, a TLS server might signal that 0-RTT keys are available.
 
-1-RTT keys are used for packets in both directions.  0-RTT keys are only
-used to protect packets sent by the client.
+Note that although TLS only uses one encryption level at a time, QUIC
+may use more than one level. For instance, after sending its Finished
+message (using a CRYPTO_HS frame in Handshake encryption) may send STREAM
+data (in 1-RTT encryption). However, if the Finished is lost, the client
+would have to retransmit the Finished, in which case it would use
+Handshake encryption.
 
 
-### Secret Export
-
-Details how secrets are exported from TLS are included in {{key-expansion}}.
-
 
 ### TLS Interface Summary
 
 {{exchange-summary}} summarizes the exchange between QUIC and TLS for both
-client and server.
+client and server. Each arrow is tagged with the encryption level used for
+that transmission.
 
 ~~~
 Client                                                    Server
 
 Get Handshake
-0-RTT Key Ready
-                      --- send/receive --->
+                      Initial ------------>
+Rekey out to 0-RTT Keys
```

Comment:
"Rekey out" and "Rekey in" don't immediately communicate that "out" and "in" are the keys in use being updated; they sound like compound verbs ("take out," "bring in").  Maybe use something like Tx and Rx to avoid this confusion?

Reply:
Done